### PR TITLE
Update numpy version requirement reflecting numpy C API change

### PR DESCRIPTION
**Pull request by fzeiser**
Trying to see whether this fixes issue #178, as indicated by @vetlewi comment and also found on [stackoverflow]( https://stackoverflow.com/questions/66060487/valueerror-numpy-ndarray-size-changed-may-indicate-binary-incompatibility-exp).

Probably we should just require people to keep the same numpy version (require `>=1.20` for all systems); I mostly wanted to see whether I understand the syntax correctly :).

### CODE REVIEW

**Comment by fzeiser** [comment]
@vetlewi: The problem seems fixed now. Would you suggest keeping different requirements for MacOs and linux with regards to the `numpy` version, or creating a common requirement? Once one thinks of `install_requires` as hard requirements, I somehow think that the current solution might be preferred over a hard requirements.

I additionally added a fully version pinned `requirements.txt` (reading more, it seems rather like a *I'm sure this works* file, than strict requirements).

**Comment by vetlewi** [comment]
> @vetlewi: The problem seems fixed now. Would you suggest keeping different requirements for MacOs and linux with regards to the `numpy` version, or creating a common requirement? Once one thinks of `install_requires` as hard requirements, I somehow think that the current solution might be preferred over a hard requirements.
> 
> I additionally added a fully version pinned `requirements.txt` (reading more, it seems rather like a _I'm sure this works_ file, than strict requirements).

I think that the reason for the failure was that a newer version of NumPy was used during build than the installed version. The only reason that numpy was added as a setup requirement in `pyproject.toml` was because `setup.py` uses `numpy.loadtxt()`. If you move the list of requirements to the setup script, you should be able to remove the numpy requirement from the `pyproject.toml`. This way we won't need to specify specific versions of packages (me being afraid that this might brake other packages in the users python environment).

**Comment by fzeiser** [comment]
> This way we won't need to specify specific versions of packages (me being afraid that this might brake other packages in the users python environment).

We'll still need the `numpy>=1.17` due to the way the random number generator is used.

**Comment by fzeiser** [comment]
... and it seems like `numpy` is still used in the `setup.py` file:

https://github.com/oslocyclotronlab/ompy/blob/09a91ffb720dac7c5fc75110b93dd13f31918913/setup.py#L143-L153

**Comment by vetlewi** [comment]
> ... and it seems like `numpy` is still used in the `setup.py` file:
> 
> https://github.com/oslocyclotronlab/ompy/blob/09a91ffb720dac7c5fc75110b93dd13f31918913/setup.py#L143-L153

The cause of the problem is because `pip` will set up a sandbox environment while building ompy (compiling cython and C++ sources). While setting up this build environment it will pull the latest versions of the packages given in `pyproject.toml` and since pip doesn't automatically upgrade packages the version of numpy during build will be different from the runtime version.
I think the best solution would be to set the requirement `numpy>=1.20` since this is where there is a breaking change occurs.

**Comment by vetlewi** [comment]
Btw... The error occurs on both Mac and linux, so there isn't any reason to use different requirements

**Comment by fzeiser** [comment]
> I think the best solution would be to set the requirement `numpy>=1.20` since this is where there is a breaking change occurs.

Where would you add this requirement? This is, I think, what I tried in 67bd4bf

**Comment by fzeiser** [comment]
> Btw... The error occurs on both Mac and linux, so there isn't any reason to use different requirements

Hm? I did not get an error and as far as I can see, travis ran on linux, too.

**Comment by vetlewi** [comment]
> > Btw... The error occurs on both Mac and linux, so there isn't any reason to use different requirements
> 
> Hm? I did not get an error and as far as I can see, travis ran on linux, too. See e.g. the build of #179: https://travis-ci.com/github/oslocyclotronlab/ompy/jobs/480616475

Hannah got the error a few weeks ago on her machine.

**Comment by fzeiser** [comment]
> > > Btw... The error occurs on both Mac and linux, so there isn't any reason to use different requirements
> > 
> > 
> > Hm? I did not get an error and as far as I can see, travis ran on linux, too. See e.g. the build of #179: https://travis-ci.com/github/oslocyclotronlab/ompy/jobs/480616475
> 
> Hannah got the error a few weeks ago on her machine.

Weird. It'd have been nice if she had reported it... From Travis and my own machine I couldn't have guessed. From your answer I strongly assume that the had the exact same error, right?

**Comment by vetlewi** [comment]
> > > > Btw... The error occurs on both Mac and linux, so there isn't any reason to use different requirements
> > > 
> > > 
> > > Hm? I did not get an error and as far as I can see, travis ran on linux, too. See e.g. the build of #179: https://travis-ci.com/github/oslocyclotronlab/ompy/jobs/480616475
> > 
> > 
> > Hannah got the error a few weeks ago on her machine.
> 
> Weird. It'd have been nice if she had reported it... From Travis and my own machine I couldn't have guessed. From your answer I strongly assume that the had the exact same error, right?

Yes, exactly the same error. We didn't think about reporting it as I knew the solution was to upgrade numpy with pip.

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,9 @@
 [build-system]
-requires = ["setuptools", "numpy", "cython", "pybind11>=2.6.0", "wheel"]
-build-backend = "setuptools.build_meta"
\ No newline at end of file
+requires = [
+"setuptools",
+"numpy; sys_platform != 'darwin'",
+"numpy>=1.20; sys_platform == 'darwin'",
```
Comment (vetlewi) [inline pyproject.toml:5]: Maybe keep line 5 and remove `; sys_platform == 'darwin'",`

**Comment by fzeiser** [comment]
> While setting up this build environment it will pull the latest versions of the packages given in `pyproject.toml` and since pip doesn't automatically upgrade packages the version of numpy during build and runtime will be different.

Anyhow, let's get back to trying to find a good solution. I don't quite understand that this is what is going on. Because I checked now: My pip knows that `numpy 1.20.1` exists; I have `numpy==1.19.2` installed. During the build and installation it seems just fine:
> Requirement already satisfied: numpy>=1.17; sys_platform != "darwin" in /home/fabiobz/miniconda3/envs/myenv/lib/python3.7/site-packages (from OMpy==1.1.0.dev0+e5cc79b) (1.19.2)

Then it proceeds with the installation.

If we just set a higher version requirement for `numpy`, shouldn't/won't this problem just repeat next time a person has installed e.g. `numpy=1.20` but `numpy=1.21` was downloaded and used for the installation?

File: setup.py [hunk 2]
```diff
@@ -160,7 +160,21 @@ def write_version_py(filename='ompy/version_setup.py'):
                                               "-march=native"])
 ]
 
-install_requires = numpy.loadtxt("requirements.txt", dtype="str").tolist()
+install_requires = [
+ "cython",
+ "numpy>=1.17; sys_platform != 'darwin'",
+ "numpy>=1.20; sys_platform == 'darwin'",
```
Comment (vetlewi) [inline setup.py:166]: Same as comment in `pyproject.toml`

**Review by vetlewi — COMMENTED** [review]
Small change, but then I guess it should work :)

File: pyproject.toml (hunk 1 above)
Comment (fzeiser) [reply pyproject.toml:5]: Waiting for your reply on the comment above: Still, I don't understand why we should require `numpy=1.20` (for everyone) if it obviously worked with `1.17`and `1.19` *before* and no changes to the code here have been made.

Comment (vetlewi) [reply pyproject.toml:5]: Either both has to be `numpy<1.20` or `numpy>=1.20`.

Comment (fzeiser) [reply pyproject.toml:5]: And I wonder how we may specify exactly this?

Comment (fzeiser) [reply pyproject.toml:5]: Maybe I'm just too dumb to get the details; shall we have a short meeting on it later? Might be quicker to resolve the misunderstanding that way.

**Comment by fzeiser** [comment]
As Vetle pointed out to me in a short meeting, this is due to the [C API change](https://numpy.org/doc/stable/release/1.20.0-notes.html#size-of-np-ndarray-and-np-void-changed) in `numpy 1.20.0`.

There is also a detailed discussion of potential solutions in:
https://github.com/scikit-learn-contrib/hdbscan/issues/457#issuecomment-773671043

In short, the two solutions I would prefer:
1) require `numpy>=1.20`
2) use `oldest-supported-numpy` in the `pyproject.toml` [not 100% sure, but should work according to the issue; sandboxed-building with an older version works, but not with a newer version than the installed version]
    
The only downside of solution 1 is that it might cause issues with other packages. Let's give it a try; people may still find back to this issue if needed.

**Comment by vetlewi** [comment]
> As Vetle pointed out to me in a short meeting, this is due to the [C API change](https://numpy.org/doc/stable/release/1.20.0-notes.html#size-of-np-ndarray-and-np-void-changed) in `numpy 1.20.0`.
> 
> There is also a detailed discussion of potential solutions in:
> [scikit-learn-contrib/hdbscan#457 (comment)](https://github.com/scikit-learn-contrib/hdbscan/issues/457#issuecomment-773671043)
> 
> In short, the two solutions I would prefer:
> 
> 1. require `numpy>=1.20`
> 2. use `oldest-supported-numpy` in the `pyproject.toml` [not 100% sure, but should work according to the issue; sandboxed-building with an older version works, but not with a newer version than the installed version]
> 
> The only downside of solution 1 is that it might cause issues with other packages. Let's give it a try; people may still find back to this issue if needed.

Nr. 2 is an interesting solution, but I will think that solution 1 is simpler and will probably work. The only packages that may break are packages that build binary files that links to numpy.

Anyways I think that at one point it will make sense to educate users about virtualenv & pyenv and the wonders of multiple python installations without conflicts!

**Pull request merged**